google/cloud/storage/oauth2/google_credentials.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/oauth2/google_credentials.h"
+
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/oauth2/credentials.h"
+#include "google/cloud/storage/version.h"
+#include <cstdlib>
+#include <fstream>
+#include <memory>
+#include <sstream>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace oauth2 {
+
+std::shared_ptr<Credentials> GoogleDefaultCredentials() {
+  std::string path = internal::GoogleAdcFilePathOrEmpty();
+  if (not path.empty()) {
+    std::ifstream is(path);
+    std::string contents(std::istreambuf_iterator<char>{is}, {});
+
+    auto cred_json = storage::internal::nl::json::parse(contents);
+    std::string cred_type = cred_json.value("type", "no type given");
+    if (cred_type == "authorized_user") {
+      return std::make_shared<AuthorizedUserCredentials<>>(contents);
+    }
+    if (cred_type == "service_account") {
+      return std::make_shared<ServiceAccountCredentials<>>(contents);
+    }
+    google::cloud::internal::RaiseRuntimeError(
+        "Unsupported credential type (" + cred_type +
+        ") when reading "
+        "Application Default Credentials file.");
+  } else {  // Check for implicit environment-based credentials.
```

Comment:
nit-ish: no `else` if we return or raise in the `if()` block then else is just adding indentation for no value.